File: go/core/interfaces/adaptive_throttler.go
package interfaces

import "context"

// AdaptiveThrottler applies client-side adaptive throttling (Google SRE "Handling Overload"):
// it tracks the recent accept/reject ratio and, when a backend starts failing, rejects a
// growing fraction of outbound requests locally — shedding load before it reaches the
// struggling backend and easing off as the backend recovers. It composes with the retry
// budget: a locally-throttled request is a terminal rejection, not retried.
//
// A peer resilience contract of Retrier and Bulkhead: it wraps an operation with a policy and
// depends on no concrete implementation. Swap implementations via the adaptivethrottle.New
// factory.
type AdaptiveThrottler interface {
	// Do runs op unless the throttler decides to reject it locally, in which case it returns a
	// coded "throttled" rejection (see adaptivethrottle) WITHOUT invoking op. Each op outcome
	// (accepted/rejected by the backend) feeds the accept-rate window that drives the decision.
	Do(ctx context.Context, op func() error) error
}

File: python/techai_webutils/src/techai_webutils/core/interfaces/adaptive_throttler.py
"""Client-side adaptive-throttling interface.

Mirrors Go's ``interfaces.AdaptiveThrottler``. An ``AdaptiveThrottler`` rejects a growing
fraction of outbound requests locally when a backend's accept-rate drops (Google SRE
"Handling Overload"), shedding load before it reaches the struggling backend.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Protocol, runtime_checkable

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable


@runtime_checkable
class AdaptiveThrottler(Protocol):
    """Sheds outbound load client-side when a backend starts failing.

    A peer resilience contract of ``Retrier``/``Bulkhead``: it wraps an operation with a policy
    and depends on no concrete implementation. Composes with the retry budget -- a locally
    throttled request is a terminal rejection, not retried.
    """

    async def do[T](self, op: Callable[[], Awaitable[T]]) -> T:
        """Run ``op`` unless the throttler rejects it locally.

        On local rejection, raises a coded "throttled" error WITHOUT invoking ``op``; otherwise
        ``op``'s outcome feeds the accept-rate window that drives the decision.
        """
        ...
